_posts/2020-03-04-eda-architecture2.md

```diff
@@ -0,0 +1,73 @@
+---
+layout: post
+title:  "An event driven architecture — part 2"
+excerpt: "Core design principles."
+date:   2020-03-04 12:53:00 +0100
+categories: architecture
+tags:  services event bus architecture
+---
+
+In the previous post in this series, we explained why we ditched our old architecture based on synchronous REST services for a completely asynchronous event driven architecture.
+
+Today, we address the core design principles that were crucial in the success of this enterprise.
+
+## Business Services and Data Processing Services
+
+We make a distinction between *Business Services* and *Data Processing Services* (a.k.a. utility services) to cleanly separate business logic from data processing complexity.
+
+DP Services are expected to be  **pure, stateless, services** that provide some kind of algorithmic data processing (computations, transformations, …). Moreover, they are also context free: they should not depend on business rules, assumptions or external data sources. All they need to do their processing must be in the message they receive. They should not have to query a tier to get more data. DP services are kind of *universal* libraries and can even be provided by tiers.
+
+Business Services implement the customers' workflows and only focus on business rules and requirements to orchestrate and implement the value addition upon our customers' audio and data. They make use of the DP services as a library for that. There are very specific to us: you would never want to externalize your business services.
+
+Business Services persist the data they produce and are their unique trusted source of truth.
+
+Business Services build and maintain their own customer configuration from events on the bus.
+
+All services must be idempotent, that is, if they receive twice the exactly same message, they must behave identically and produce the same outputs.
+
+## Commands, Results and Events
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages published on the bus by business services and telling the world what happened.
+
+A business service owns the type of Events it emits. It knows nothing about the services who will process them. It subscribes to the types of Events it needs but knows nothing about their origins.
+
+An Event type defines the meaning of the events of that type and their data schema. They must be documented.
+
+The type of an actual event message is given by its name (a.k.a. *routing key* because it is used by the subscription routing). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that it is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
```

Comment:
I don't know the rule regarding ellipsis following a comma or not, but you've written `(computations, transformations, …)` above, but ``ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`…` here.
We might consider formatting with the same rule.